README
Crypt-FNA

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Crypt::FNA

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-FNA

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Crypt-FNA

    CPAN Ratings
        http://cpanratings.perl.org/d/Crypt-FNA

    Search CPAN
        http://search.cpan.org/dist/Crypt-FNA/
        
    Algorithm description
    	http://www.perl.it/documenti/articoli/2010/04/anakryptfna.html
  


ERROR CODE

	0 "Order of the curve is not correct.Must necessarily be numeric. Ex. r=7"
	1 "Order of the curve must be a number greater than 0"
	2 "Length Square container is incorrect. Must necessarily be numeric"
	3 "Side of a square container fractal must be a number greater than 0"
	5"Value of is not correct. Must necessarily be numeric.Default loaded"
	6 "The angle must be expressed in the system sessadecimal (ex. 126.35)Default loaded"
	7 "Error reading sub encrypt, package anakrypt: FNA"
	8 "error writing file, package anakrypt: FNA sub encrypt"
	9 "read error on sub decrypt myInput package Anakrypt: FNA"
	10 "write error on sub decrypt MYOUTPUT package Anakrypt: FNA"
	11 "error writing PNG sub draw_fract package Anakrypt: FNA"
	12 "error background: only numeric character (RGB)"
	13 "error background: only three number (RGB) from 0 to 255"
	14 "error foreground: only numeric character (RGB)"
	15 "error foreground: only three number (RGB) from 0 to 255"
	16 "error loading GD::Simple, drawing aborted"
	18 "error zoom: the value must be a number greater than zero."
	19 "errors during object instantiation."
	20 "error magic setting."
	21 "error 'salted' setting (is a boolean, only 'true' or 'false')."
	22 "error loading Tie::File."
	23 "Error reading sub mac, package anakrypt: FNA"	

COPYRIGHT AND LICENCE

Copyright (C) 2009 2010 Mario Rossano aka Anak
birthday 05/08/1970; birthplace: Italy

This program is free software; you can redistribute it and/or modify it
under the terms of either:
CC-NC-BY-SA 
license http://creativecommons.org/licenses/by-nc-sa/2.5/it/deed.en
Creative Commons License: http://i.creativecommons.org/l/by-nc-sa/2.5/it/88x31.png
FNA Fractal Numerical Algorithm for a new cryptography technology, author Mario Rossano
is licensed under a: http://creativecommons.org/B/by-nc-sa/2.5/it/deed.en - Creative Commons Attribution-Noncommercial-Share Alike 2.5 Italy License
Permissions beyond the scope of this license may be available at software@netlogicalab.com




Changes
Revision history for Crypt-FNA

0.64    18/06/2013 20:45
	fixed crypt_fract method

0.63    03/10/2012 17:49
	fixed zoom (if != 1) on make_fract method

0.62    30/10/2012 17:00
	fixed abstract

0.61    29/10/2012 16:05
	fixed zoom factor on method make_fract
	
0.60    29/10/2012 15:05
	fixed error on makefile.pl

0.59    29/10/2012 11:32
	rebuilt distribution module with module-starter

0.58    26/10/2012 10:40
	code optimization

0.56    26/03/2012 08:13
	Fixed little bug make_fract method

0.54    15/08/2011 13:32
	Added 'MAC' method (digest)

0.48    16/05/2011 12:03
	RAM usage optimization

0.24    03/12/2010 17:01
	Added 'salted' attribute
	Added 'make_salt' - internal - method
	Modified POD text
	Added control on validation.pm (rev. 0.02) to verify attribute 'salted' is a boolean
	Updated array of message error (code 21): salted initialization  error (only true | false)

0.12    27/09/2010 13:30
	Modified s/GD::SIMPLE/GD::Simple/ 

0.10    26/07/2010 16:13
	Added method 'decrypt_scalar'
	Added POD information for 'decrypt_scalar' method

0.06    15/04/2010 11:05
	Added POD information for 'new' method of Crypt::FNA

0.05    14/04/2010 11:05
	Add more information to Crypt::FNA POD
	Add more information to Crypt::FNA::Validation POD
	Modified 'Changes' (last version on top)

0.04    09/04/2010 12:55
	Modified abstract in build.pl
        Correct text of readme

0.03    09/04/2010 12:55
        Correct text of POD Crypt::FNA
        Add POD to Crypt::FNA::Validation
        Add Version to Crypt::FNA::Validation
        Add t/01-add.t
        
0.02    09/04/2010 9:55
        Correct text of POD
        
0.01    16/03/2010/23:26
        First version, released on an unsuspecting world.


lib/Crypt/FNA/Validation.pm
# package: Anak Cryptography with Fractal Numerical Algorithm FNA
# author: Mario Rossano aka Anak, www.netlogicalab.com, www.netlogica.it; software@netlogicalab.com; software@netlogica.it
# birthday 05/08/1970; birthplace: Italy
# Validation FILE

# Copyright (C) 2009 Mario Rossano aka Anak
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of either:
# CC-NC-BY-SA 
# license http://creativecommons.org/licenses/by-nc-sa/2.5/it/deed.en
# Creative Commons License: http://i.creativecommons.org/l/by-nc-sa/2.5/it/88x31.png
# FNA Fractal Numerical Algorithm for a new cryptography technology, author Mario Rossano
# is licensed under a: http://creativecommons.org/B/by-nc-sa/2.5/it/deed.en - Creative Commons Attribution-Noncommercial-Share Alike 2.5 Italy License
# Permissions beyond the scope of this license may be available at software@netlogicalab.com

package Crypt::FNA::Validation;

	use strict;
	use warnings;
	our $VERSION =  '0.01';

	my $fna_default_values= (
		{
			r => 7,
			angle => [0,60,-60,0],
			square => 4096,
			background => [255,255,255],
			foreground => [0,0,0],
			magic => 3,
			zoom => 2
		}
	);
	
	sub new {
		my $class = shift;
		my $init  = shift;
		my $self={};

		bless $self,$class;
		
		$self->intercept($init->{intercept});

		return $self	
	}
	
	sub intercept {
		my $self=shift;
		if (@_) {
			$self->{intercept}=shift
		}
		return $self->{intercept}
	}

	sub method_new_fna {
		my $self=shift;
		my @error_messages="";
		@{$self->intercept->{message}}=();
		
		$self->intercept->{r}.="";
		if ($self->intercept->r eq "" | $self->intercept->r eq "0") {
			$self->intercept->{r}=$fna_default_values->{r}
		} else {
			if ($self->intercept->r=~/[^0-9.]/) {
				push(@error_messages,0);
				$self->intercept->{r}=$fna_default_values->{r}
			} 
		}

		if (!$self->intercept->angle) {
			@{$self->intercept->{angle}}=@{$fna_default_values->{angle}}
		} else {
			for (my $k=0;$k<scalar(@{$self->intercept->angle});$k++) {
				if (${$self->intercept->angle}[$k]=~/[^0-9.+-]/) {
					push(@error_messages,5);
					@{$self->intercept->angle}=@{$fna_default_values->{angle}};
					last
				}
			}
		}

		$self->intercept->{square}.="";
		if ($self->intercept->square eq "" | $self->intercept->square eq "0") {
			$self->intercept->{square}=$fna_default_values->{square}
		} else {
			if ($self->intercept->square=~/[^0-9.]/) {
				push(@error_messages,2);
				$self->intercept->{square}=$fna_default_values->{square}
			}
		}

		if (!$self->intercept->background) {
			@{$self->intercept->{background}}=@{$fna_default_values->{background}}
		} else {
			 if (scalar(@{$self->intercept->background})>3) {
				push(@error_messages,13);
				@{$self->intercept->background}=@{$fna_default_values->{background}}
			 } else {
				for (my $k=0;$k<scalar(@{$self->intercept->background});$k++) {
					if (${$self->intercept->background}[$k]=~/[^0-9]/) {
						push(@error_messages,12);
						@{$self->intercept->background}=@{$fna_default_values->{background}}
					} else {
						if (${$self->intercept->background}[$k]>255) {
							push(@error_messages,13);
							@{$self->intercept->background}=@{$fna_default_values->{background}}
						}
					}
				}
			}
		}

		if (!$self->intercept->foreground) {
			@{$self->intercept->{foreground}}=@{$fna_default_values->{foreground}}
		} else {
			if (scalar(@{$self->intercept->foreground})>3) {
				push(@error_messages,15);
				@{$self->intercept->foreground}=@{$fna_default_values->{foreground}}
			} else {
				for (my $k=0;$k<scalar(@{$self->intercept->foreground});$k++) {
					if (${$self->intercept->foreground}[$k]=~/[^0-9]/) {
						push(@error_messages,14);
						@{$self->intercept->foreground}=@{$fna_default_values->{foreground}}
					} else {
						if (${$self->intercept->foreground}[$k]>255) {
							push(@error_messages,15);
							@{$self->intercept->foreground}=@{$fna_default_values->{foreground}}
						}
					}
				}
			}
		}

		$self->intercept->{magic}.="";
		if ($self->intercept->magic eq "") {
			$self->intercept->{magic}=$fna_default_values->{magic}
		} else {
			if ($self->intercept->magic=~/[^0-9]/) {
				push(@error_messages,20);
				$self->intercept->{magic}=$fna_default_values->{magic}
			}
		}
		
		@{$self->intercept->message}=@error_messages;
		@error_messages="";
		return @{$self->intercept->message}
	}

	sub param_zoom_fna {
		my $self=shift;

		if (${$self->intercept}[0] =~ /[^0-9.]/) {
			push(@{$self->intercept->[1]->{message}},18);
			${$self->intercept}[0]=$fna_default_values->{zoom}
   		}
  		return (${$self->intercept}[0],@{@{$self->intercept}[1]->{message}})
		
	}

	sub open_file {
		my $self=shift;
		
		push(@{$self->intercept->[1]->{message}},${$self->intercept}[0]);
		return @{@{$self->intercept}[1]->{message}}
	}
1;

# POD SECTION

=head1 NAME

Crypt::FNA::Validation


=head1 VERSION

Version 0.01


=head1 DESCRIPTION

Crypt::FNA::Validation is a service package for FNA object validation in FNA.pm


=head1 METHODS


=head2 new

Make a 'validation' object (used in fna.pm)


=head2 intercept

Intercept and validate Crypt::FNA attribute or method


=head2 method_new_fna

Validate Crypt::FNA 'new' method


=head2 open_file

Validate Crypt::FNA 'open' instruction


=head2 param_zoom_fna

Validate Crypt::FNA param 'zoom' for drawing fractal curve


=head1 AUTHOR

  Mario Rossano
  software@netlogicalab.com
  software@netlogica.it
  http://www.netlogicalab.com


=head1 BUGS

Please, send me your alerts to software@netlogica.it


=head1 SUPPORT

Write me :) software@netlogica.it


=head1 COPYRIGHT & LICENSE

FNA by Mario Rossano, http://www.netlogicalab.com

This pod text by Mario Rossano

Copyright (C) 2009 Mario Rossano aka Anak
birthday 05/08/1970; birthplace: Italy

This program is free software; you can redistribute it and/or modify it
under the terms of either:
CC-NC-BY-SA
license http://creativecommons.org/licenses/by-nc-sa/2.5/it/deed.en
Creative Commons License: http://i.creativecommons.org/l/by-nc-sa/2.5/it/88x31.png

FNA Fractal Numerical Algorithm for a new cryptography technology, author Mario Rossano
is licensed under a: http://creativecommons.org/B/by-nc-sa/2.5/it/deed.en - Creative Commons Attribution-Noncommercial-Share Alike 2.5 Italy License

Permissions beyond the scope of this license may be available at software@netlogicalab.com
